File: applets/src/org/rr0/ufoathome/control/AbstractController.java
package org.rr0.ufoathome.control;

import org.rr0.ufoathome.view.draw.AnimationEvent;
import org.rr0.ufoathome.view.draw.AnimationListener;
import org.rr0.ufoathome.model.ufo.MessageProducer;
import org.rr0.ufoathome.view.AbstractView;

import java.awt.*;
import java.awt.event.*;
import java.awt.image.ImageObserver;
import java.net.URL;
import java.util.*;

/**
 * @author Jerome Beau
 * @version 0.3
 */
public abstract class AbstractController extends MouseAdapter implements MouseMotionListener, KeyListener, MessageProducer {
    private Hashtable imageCache = new Hashtable();
    protected int lastX = -1;
    protected int lastY = -1;
    protected GregorianCalendar currentTime;
    protected ResourceBundle messagesBundle;
    protected Vector animationListeners = new Vector();
    protected static int lastLayersStartBit = 0;
    protected int layersStartBit = lastLayersStartBit;
    public final BitSet ALL_LAYERS = new BitSet();
    public BitSet layersToDraw = new BitSet();

    public AbstractController() {
        messagesBundle = ResourceBundle.getBundle("org.rr0.is.presentation.view.report.applet.StarSkyLabels");
    }

    public int getLayersStartBit() {
        return layersStartBit;
    }

    public Image getImage(final URL url, final Component creator) {
        Image img;
        Object o = imageCache.get(url);
        if (o != null) {
            return (Image) o;
        }
        try {
            fireMessage("LoadingImage", null);
            o = url.getContent();
            if (o == null) {
                return null;
            }
            if (o instanceof Image) {
                img = (Image) o;
            } else {
                java.awt.image.ImageProducer imageProducer = (java.awt.image.ImageProducer) o;
                img = creator.createImage(imageProducer);
                creator.prepareImage(img, new ImageObserver() {
                    private boolean loading;

                    public boolean imageUpdate(Image img, int infoflags, int x, int y, int w, int h) {
                        if ((infoflags & ImageObserver.ALLBITS) != 0) {
                            loading = false;
                        } else if (!loading && (infoflags & ImageObserver.SOMEBITS) != 0) {
                            //                            String message = messagesBundle.getString("LoadingImage");
                            //                            fireMessage(message);
                            loading = true;
                        } else if ((infoflags & ImageObserver.ERROR) != 0) {
                            System.err.println("Error while loading " + img);
                        }
                        boolean b = creator.imageUpdate(img, infoflags, x, y, w, h);
                        return b;
                    }
                });
            }

            imageCache.put(url, img);
            return img;

        } catch (Exception ex) {
            ex.printStackTrace();
            return null;
        }
    }

    public int getLastX() {
        return lastX;
    }

    public int getLastY() {
        return lastY;
    }

    public void setLastPos(int lastX, int lastY) {
        this.lastX = lastX;
        this.lastY = lastY;
    }

    public GregorianCalendar getCurrentTime() {
        return currentTime;
    }

    public void setTime(GregorianCalendar currentTime) {
        this.currentTime = currentTime;
    }

    public ResourceBundle getMessagesBundle() {
        return messagesBundle;
    }

    public void draw(BitSet requiredLayers) {
        // Clear
        for (int i = 0; i < layersToDraw.size(); i++) {
            layersToDraw.clear(i);
        }
    }

    protected void fireTimeChanged(GregorianCalendar currentTime) {
        for (int i = 0; i < animationListeners.size(); i++) {
            AnimationListener animationListener = (AnimationListener) animationListeners.elementAt(i);
            AnimationEvent animationEvent = new AnimationEvent(this, currentTime);
            animationListener.timeChanged(animationEvent);
        }
    }

    public abstract AbstractView getView();

    public void display() {
        getView().displayBuffered();
    }

    public void selectObject(String objectName) {
        // Do nothing by default
    }

    public void setZoomFactor(int x) {
        // Do nothing by default
    }

    /**
     * Invoked when left click on background (i.e. not on a specific object).
     *
     * @param e The mouse event
     */
    public void backgroundClicked(MouseEvent e) {
        // Do nothing by default
        e.consume();
    }

    public void keyPressed(KeyEvent e) {
        // Do nothing by default
        e.consume();
    }

    public void keyTyped(KeyEvent e) {
        // Do nothing by default
        e.consume();
    }

    public void keyReleased(KeyEvent e) {
        // Do nothing by default
        e.consume();
    }

    protected boolean isToDraw(BitSet layers, int someLayer) {
        boolean layerActive = layers.get(someLayer) && layersToDraw.get(someLayer);
        return layerActive;
    }

    public void draw() {
        draw(ALL_LAYERS);
        display();
    }

    protected int newLayer() {
        ALL_LAYERS.set(layersStartBit);
        return layersStartBit++;
    }

    public void modified(int layer) {
        for (int i = 0; i <= layer; i++) {
            layersToDraw.set(i);
        }
    }
}

File: applets/src/org/rr0/ufoathome/control/AbstractController.ts
import {AbstractView} from "../view/AbstractView";
import {MessageProducer} from "../model/ufo/MessageProducer";
import {BitSet} from "../BitSet";
import {GregorianCalendar} from "../GregorianCalendar";
import {ResourceBundle} from "../ResourceBundle";
import {Hashtable} from "../Hashtable";
import {Image} from "../view/gui/Image";
import {Component} from "../view/gui/Component";
import {AnimationListener} from "../view/draw/AnimationListener";
import {AnimationEvent} from "../view/draw/AnimationEvent";
import {ImageObserver} from "../view/gui/ImageObserver";
import {KeyEvent} from "../view/gui/KeyEvent";
import {MouseAdapter} from "../view/gui/MouseAdapter";

export abstract class AbstractController extends MouseAdapter implements MouseMotionListener, KeyListener, MessageProducer {
  private imageCache = new Hashtable();
  protected lastX = -1;
  protected lastY = -1;
  protected currentTime: GregorianCalendar;
  protected messagesBundle: ResourceBundle;
  protected animationListeners = [];
  protected lastLayersStartBit = 0;
  protected layersStartBit: number = this.lastLayersStartBit;
  public ALL_LAYERS = new BitSet();
  public layersToDraw = new BitSet();

  constructor() {
    super();
    this.messagesBundle = ResourceBundle.getBundle("org.rr0.is.presentation.view.report.applet.StarSkyLabels");
  }

  public getLayersStartBit() {
    return this.layersStartBit;
  }

  public getImage(url: URL, creator: Component): Image {
    let img: Image;
    let o: Object = this.imageCache.get(url);
    if (o != null) {
      return <Image>o;
    }
    try {
      this.fireMessage("LoadingImage", null);
      o = url.getContent();
      if (o == null) {
        return null;
      }
      if (o instanceof Image) {
        img = <Image> o;
      } else {
        const imageProducer = <ImageProducer>o;
        img = creator.createImage(imageProducer);
        creator.prepareImage(img, new class implements ImageObserver {
          private boolean;

          imageUpdate(img:Image, infoflags: number, x: number, y: number, w: number, h: number): boolean {
            if ((infoflags & ImageObserver.ALLBITS) != 0) {
              loading = false;
            } else if (!loading && (infoflags & ImageObserver.SOMEBITS) != 0) {
              //                            String message = messagesBundle.getString("LoadingImage");
              //                            fireMessage(message);
              loading = true;
            } else if ((infoflags & ImageObserver.ERROR) != 0) {
              console.error("Error while loading " + img);
            }
            const b = creator.imageUpdate(img, infoflags, x, y, w, h);
            return b;
          }
        });
      }

      this.imageCache.put(url, img);
      return img;

    } catch (ex) {
      ex.printStackTrace();
      return null;
    }
  }

  public getLastX() {
    return this.lastX;
  }

  public getLastY() {
    return this.lastY;
  }

  public setLastPos(lastX: number, lastY: number) {
    this.lastX = lastX;
    this.lastY = lastY;
  }

  public getCurrentTime(): GregorianCalendar {
    return this.currentTime;
  }

  public setTime(currentTime: GregorianCalendar) {
    this.currentTime = currentTime;
  }

  public getMessagesBundle(): ResourceBundle {
    return this.messagesBundle;
  }

  public draw(requiredLayers: BitSet) {
    // Clear
    for (let i = 0; i < this.numberlayersToDraw.length; i++) {
      this.layersToDraw.clear(i);
    }
  }

  protected fireTimeChanged(currentTime: GregorianCalendar) {
    for (let i = 0; i < this.animationListeners.length; i++) {
      const animationListener = <AnimationListener>this.animationListeners[i];
      const animationEvent = new AnimationEvent(this, currentTime);
      animationListener.timeChanged(animationEvent);
    }
  }

  public getView(): AbstractView {
    throw Error('Not implemented belmow')
  }

  public display() {
    this.getView().displayBuffered();
  }

  public selectObject(objectName: String) {
    // Do nothing by default
  }

  public setZoomFactor(x: number) {
    // Do nothing by default
  }

  /**
   * Invoked when left click on background (i.e. not on a specific object).
   *
   * @param e The mouse event
   */
  public backgroundClicked(e: MouseEvent) {
    // Do nothing by default
    e.preventDefault();
  }

  public keyPressed(e: KeyEvent) {
    // Do nothing by default
    e.preventDefault();
  }

  public keyTyped(e: KeyEvent) {
    // Do nothing by default
    e.preventDefault();
  }

  public keyReleased(e: KeyEvent) {
    // Do nothing by default
    e.preventDefault();
  }

  protected isToDraw(layers: BitSet, someLayer: number): boolean {
    const layerActive = layers.get(someLayer) && this.layersToDraw.get(someLayer);
    return layerActive;
  }

  public draw() {
    this.draw(this.ALL_LAYERS);
    this.display();
  }

  protected newLayer(): number {
    this.ALL_LAYERS.set(this.layersStartBit);
    return this.layersStartBit++;
  }

  public modified(layer: number) {
    for (let i = 0; i <= layer; i++) {
      this.layersToDraw.set(i);
    }
  }
}
